File: csharp/protein-translation/ProteinTranslation.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class ProteinTranslation
{

    private const int CodonLength = 3;
    public static string[] Proteins(string strand) =>
        ToAminoAcids(strand).ToArray();

    private static IEnumerable<string> ToAminoAcids(string strand)
    {
        for (var idx = 0; idx < strand.Length; idx += CodonLength)
        {
            var protein = codonToAminoAcid(strand[idx..(idx + CodonLength)]);
            if (protein == "STOP")
                yield break;
            yield return protein;
        }
    }

    private static string codonToAminoAcid(string codon) =>
        codon switch
        {
            "UCU" or "UCC" or "UCA" or "UCG" => "Serine",
            "UAA" or "UAG" or "UGA" => "STOP",
            "UUU" or "UUC" => "Phenylalanine",
            "UUA" or "UUG" => "Leucine",
            "UAU" or "UAC" => "Tyrosine",
            "UGU" or "UGC" => "Cysteine",
            "AUG" => "Methionine",
            "UGG" => "Tryptophan",
            _ => throw new ArgumentOutOfRangeException(codon),
        };
}

File: python/protein-translation/protein_translation.py
from collections.abc import Generator


CODON_SIZE = 3

AMINOACIDS = {
    "AUG": "Methionine",
    "UUU": "Phenylalanine",
    "UUC": "Phenylalanine",
    "UUA": "Leucine",
    "UUG": "Leucine",
    "UCU": "Serine",
    "UCC": "Serine",
    "UCA": "Serine",
    "UCG": "Serine",
    "UAU": "Tyrosine",
    "UAC": "Tyrosine",
    "UGU": "Cysteine",
    "UGC": "Cysteine",
    "UGG": "Tryptophan",
    "UAA": None,
    "UAG": None,
    "UGA": None,
}


def proteins(strand: str) -> list[str]:
    return list(takewhile(batch(strand)))


def batch(strand: str) -> Generator[str, None, None]:
    for i in range(0, len(strand), CODON_SIZE):
        yield strand[i : i + CODON_SIZE]


def takewhile(codons: Generator[str, None, None]) -> Generator[str, None, None]:
    for codon in codons:
        amino_acid = AMINOACIDS[codon]
        if amino_acid is None:
            break
        yield amino_acid
